packages/frontend/src/redux/slices/tokens/index.js

```diff
@@ -0,0 +1,173 @@
+        return contractMetadata;
+    }
+    return getMetadata({ contractName, accountId });
+}
+
+const fetchOwnedTokensForContract = createAsyncThunk(
+    `${SLICE_NAME}/fetchOwnedTokensForContract`,
+    async ({ accountId, contractName }, thunkAPI) => {
+        const { actions: { addTokensMetadata } } = tokensSlice;
+        const { dispatch } = thunkAPI;
+
+        const balance = await getBalanceOf({ contractName, accountId });
+
+        dispatch(addTokensMetadata({ accountId, contractName, balance }));
+    },
+    {
+        condition: ({ accountId, contractName }, thunkAPI) => {
+            const { getState } = thunkAPI;
+            if (selectOneTokenLoading(getState(), { accountId, contractName })) {
+                return false;
+            }
+        }
+    }
+);
+
+const fetchTokens = createAsyncThunk(
+    `${SLICE_NAME}/fetchTokens`,
+    async ({ accountId }, thunkAPI) => {
+        const { dispatch, getState } = thunkAPI;
+
+        const likelyContracts = [...new Set([...(await getLikelyTokenContracts({ accountId })), ...WHITELISTED_CONTRACTS])];
+
+        await Promise.all(likelyContracts.map(async contractName => {
+            const { actions: { setContractMetadata } } = tokensSlice;
+            try {
+                const contractMetadata = await getCachedContractMetadataOrFetch(contractName, accountId, getState());
+                if (!selectOneContractMetadata(getState(), { contractName })) {
+                    dispatch(setContractMetadata({ contractName, metadata: contractMetadata }));
+                }
+                await dispatch(fetchOwnedTokensForContract({ accountId, contractName, contractMetadata }));
+            } catch (e) {
+                // Continue loading other likely contracts on failures
+                console.warn(`Failed to load FT for ${contractName}`, e);
+            }
+        }));
+    }
+);
+
+const tokensSlice = createSlice({
+    name: SLICE_NAME,
+    initialState,
+    reducers: {
+        setContractMetadata(state, { payload }) {
+            const { metadata, contractName } = payload;
+            set(state, ['metadata', 'byContractName', contractName], metadata);
+        },
+        addTokensMetadata(state, { payload }) {
+            const { contractName, balance, accountId } = payload;
+            set(state, ['ownedTokens', 'byAccountId', accountId, contractName, 'balance'], balance);
+        },
+    },
+    extraReducers: ((builder) => {
+        builder.addCase(fetchOwnedTokensForContract.pending, (state, { meta }) => {
+            const { accountId, contractName } = meta.arg;
+            set(state, ['ownedTokens', 'byAccountId', accountId, contractName, 'loading'], true);
+        });
+        builder.addCase(fetchOwnedTokensForContract.fulfilled, (state, { meta }) => {
+            const { accountId, contractName } = meta.arg;
+            set(state, ['ownedTokens', 'byAccountId', accountId, contractName, 'loading'], false);
+        });
```

Comment:
Should we be re-setting `error` value in either of these?  I would assume we'd at least want to in `.fulfilled`?

Reply:
Done.